store/db.go

```diff
@@ -49,15 +51,17 @@ func newDBLock(dirPath string) (*dbLock, error) {
 }
 
 func (dl *dbLock) lock() error {
+	dl.Lock()
 	if err := dl.fl.ExclusiveLock(); err != nil {
+		dl.Unlock()
 		return err
 	}
-	dl.Lock()
 	return nil
 }
 
 func (dl *dbLock) unlock() error {
 	if err := dl.fl.Unlock(); err != nil {
+		dl.Unlock()
 		return err
 	}
 	dl.Unlock()
```

Comment:
> why not just call dl.Unlock() at the top of the function?

If your question is why I inverted the order putting the sync.Mutex lock before the flock that's because the sync.Mutex must protect the flock. The reason is that calling flock.ExclusiveLock after a previous ExclusiveLock will continue since it's on the same filedescriptor. This can cause something like this:

* Process A, goroutine 1: takes the the flock, then takes the lock on the mutex, and do db.Open
* Process A, goroutine 2: takes the flock, blocks on the lock on the mutex
* Process A, goroutine 1: does something on db, calls db.Close, release the lock on the mutex, unlocks the flock
* Process A, goroutine 2: takes the lock on the mutex, but now it has the flock UNLOCKED
* Process B, goroutine 1: take the lock on the mutex (since it's another process), and takes the lock on the flock: the ql camlistore lock will fail.

So the sync.Mutex that's really (per store instance) exclusive must go before the flock

Obviously there must be only one store instance (on the same dir) per process. This is what actually rkt does (I haven't found wrong store usage).